README.md
# music-pitch

[![Build Status](https://travis-ci.org/danigb/music-pitch.svg?branch=master)](https://travis-ci.org/danigb/music-pitch)
[![Code Climate](https://codeclimate.com/github/danigb/music-pitch/badges/gpa.svg)](https://codeclimate.com/github/danigb/music-pitch)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)
[![npm version](https://badge.fury.io/js/music-pitch.svg)](https://badge.fury.io/js/music-pitch)


`music-pitch` is the bridge between midi, your app and your synthetizers. It's a small (2.5kb minified) and fast library to manipulate note names, midi notes and note frequencies.

```js
var pitch = require('music-pitch')
// get a note from midi
var note = pitch.fromMidi( ... )
// write it to the console
console.log(note)
// give it to your synth
synth.play(pitch.toFreq(note))
```

## Installation

#### For node

Install via npm: `npm install --save music-pitch` and require it.

#### Browsers

No distribution (yet). Use webpack, browserify or a similar tool.

## Usage

#### Note names

The `str` function returns the [scientific notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation) of a given note if valid. Can be used to check if its a valid note:

```js
pitch.str('bbb') // => 'Bbb'
pitch.str('fx4') // => 'F##4'

if (pitch.str(str) !== null) { /* valid pitch str */ }
```

You can get also pitch classes (pitches without octaves), note letter, octave and accidentals:

```js
pitch.pitchClass('c##5') // => 'C##'
pitch.letter('eb3') // => 'E'
pitch.octave('eb3') // => 3
pitch.accidentals('eb3') // => 'b'
```

#### Working with midi

You have two functions for converting from and to midi numbers:

```js
pitch.toMidi('A4') // => '69'
pitch.fromMidi(69) // => 'A4'
```

#### Working with frequencies

The same way, you have two frequency related functions:

```js
pitch.toFreq('A4') // => 440
pitch.fromFreq(440) // => 'A4'
```

#### Using different pitch notation

In the case scientific notation is not what you need, you can always use [pitch array notation](https://github.com/danigb/a-pitch) for every function that expects a string:

```js
pitch.toFreq([5, 0, 3]) // => 220
```

Also you can covert from scientific notation to pitch array notation with the pitch function:

```js
pitch('A3') // => [5, 0, 3]
```

#### More...

That's all for this library, but maybe you need:

- [pitch-transpose](https://github.com/danigb/pitch-transpose)
- [music-interval](https://github.com/danigb/music-interval)
- [music-scale](https://github.com/danigb/music-scale)

## API

<!-- START docme generated API please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN docme TO UPDATE -->

<div>
<div class="jsdoc-githubify">
<section>
<article>
<div class="container-overview">
<dl class="details">
</dl>
</div>
<dl>
<dt>
<h4 class="name" id="accidentals"><span class="type-signature"></span>accidentals<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the accidentals from a pitch</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L101">lineno 101</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the pitch accidentals</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.accidentals('C##3') // => '##'
pitch.accidentals('Bb4') // => 'b'
pitch.accidentals('E') // => ''</code></pre>
</dd>
<dt>
<h4 class="name" id="cents"><span class="type-signature"></span>cents<span class="signature">(from, to, decimals)</span><span class="type-signature"> &rarr; {Integer}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the distance in cents between pitches or frequencies</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>from</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Integer</span>
</td>
<td class="description last"><p>first pitch or frequency</p></td>
</tr>
<tr>
<td class="name"><code>to</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Integer</span>
</td>
<td class="description last"><p>other pitch or frequency</p></td>
</tr>
<tr>
<td class="name"><code>decimals</code></td>
<td class="type">
<span class="param-type">Integer</span>
</td>
<td class="description last"><p>the decimal precision (2 by default)</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L201">lineno 201</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the distance in cents</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">Integer</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>cents(440, 444) // => 15.66
cents('A4', 444) // => 15.66
cents('A4', 'A#4') // => 100</code></pre>
</dd>
<dt>
<h4 class="name" id="fromFreq"><span class="type-signature"></span>fromFreq<span class="signature">(freq)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the pitch of a given frequency.</p>
<p>It will round the frequency to the nearest pitch frequency. Use <code>cents</code> function
if you need to know the difference between the the frequency and the pitch.</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>freq</code></td>
<td class="type">
<span class="param-type">Float</span>
</td>
<td class="description last"><p>the frequency</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L158">lineno 158</a>
</li>
</ul></dd>
<dt class="tag-see">See:</dt>
<dd class="tag-see">
<ul>
<li><a href="global.html#cents">cents</a></li>
</ul>
</dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the pitch</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.fromFreq(440) // => 'A4'
pitch.fromFreq(443) // => 'A4'
pitch.cents(443, 'A4') // => ... to get the difference</code></pre>
</dd>
<dt>
<h4 class="name" id="fromMidi"><span class="type-signature"></span>fromMidi<span class="signature">(midi)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the pitch of the given midi number</p>
<p>This method doesn't take into account diatonic spelling. Always the same
pitch class is given to the same midi number.</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>midi</code></td>
<td class="type">
<span class="param-type">Integer</span>
</td>
<td class="description last"><p>the midi number</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L119">lineno 119</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the pitch</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.fromMidi(69) // => 'A4'</code></pre>
</dd>
<dt>
<h4 class="name" id="letter"><span class="type-signature"></span>letter<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get letter of the pitch (in uppercase)</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch as string or array</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L51">lineno 51</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the letter of the pitch in uppercase</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.letter('fx') // => 'F'</code></pre>
</dd>
<dt>
<h4 class="name" id="octave"><span class="type-signature"></span>octave<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {Integer}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the octave of a pitch</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L66">lineno 66</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the octave number</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">Integer</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.octave('F#3') // => 3</code></pre>
</dd>
<dt>
<h4 class="name" id="pitchClass"><span class="type-signature"></span>pitchClass<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the pitch class (pitch name without octaves) from a pitch</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch to get the pitchClass number from</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L82">lineno 82</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the pitch class</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.pitchClass('a4') // => 'A'
pitch.pitchClass('ab') // => 'Ab'
pitch.pitchClass('cx2') // => 'C##'</code></pre>
</dd>
<dt>
<h4 class="name" id="str"><span class="type-signature"></span>str<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {String}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get pitch name in scientific notation</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch string or array</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L36">lineno 36</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the name of the pitch</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">String</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.str('cb') // => 'Cb'
pitch.str('fx2') // => 'F##2'</code></pre>
</dd>
<dt>
<h4 class="name" id="toFreq"><span class="type-signature"></span>toFreq<span class="signature">(pitch, tuning)</span><span class="type-signature"> &rarr; {Float}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the pitch frequency in hertzs</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
</td>
<td class="description last"><p>the pitch</p></td>
</tr>
<tr>
<td class="name"><code>tuning</code></td>
<td class="type">
<span class="param-type">Integer</span>
</td>
<td class="description last"><p>optional tuning, 440 by default</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L179">lineno 179</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<ul>
<li>the pitch frequency</li>
</ul>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">Float</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.toFreq('A4') // => 440
pitch.toFreq('A3', 444) // => 222</code></pre>
</dd>
<dt>
<h4 class="name" id="toMidi"><span class="type-signature"></span>toMidi<span class="signature">(pitch)</span><span class="type-signature"> &rarr; {Integer}</span></h4>
</dt>
<dd>
<div class="description">
<p>Get the midi number of a pitch</p>
</div>
<h5>Parameters:</h5>
<table class="params">
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th class="last">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td class="name"><code>pitch</code></td>
<td class="type">
<span class="param-type">String</span>
|
<span class="param-type">Array</span>
</td>
<td class="description last"><p>the pitch string (or pitch array)</p></td>
</tr>
</tbody>
</table>
<dl class="details">
<dt class="tag-source">Source:</dt>
<dd class="tag-source"><ul class="dummy">
<li>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js">index.js</a>
<span>, </span>
<a href="https://github.com/danigb/music-pitch/blob/master/index.js#L136">lineno 136</a>
</li>
</ul></dd>
</dl>
<h5>Returns:</h5>
<div class="param-desc">
<p>the midi number</p>
</div>
<dl>
<dt>
Type
</dt>
<dd>
<span class="param-type">Integer</span>
</dd>
</dl>
<h5>Example</h5>
<pre class="prettyprint"><code>pitch.toMidi('A4') // => 69
pitch.toMidi('A3') // => 57</code></pre>
</dd>
</dl>
</article>
</section>
</div>

*generated with [docme](https://github.com/thlorenz/docme)*
</div>
<!-- END docme generated API please keep comment here to allow auto update -->

## License

MIT License


CHANGELOG
CHANGELOG

[Master]
- Add accidentals function
- Rename name to str
- Add name, letter, octave, pitchClass
- Add choma function
- Add fromFreq, toFreq and cents
- Add fromMidi and toMidi


test/tests.js
var vows = require('vows')
var assert = require('assert')
var pitch = require('../')

vows.describe('music-pitch').addBatch({
  'pitch main function': function () {
    assert.deepEqual(pitch('C#4'), [0, 1, 4])
  },
  'pitch str': {
    'str': function () {
      assert.equal(pitch.str('c'), 'C')
      assert.equal(pitch.str('bbb3'), 'Bbb3')
      assert.equal(pitch.str('fx'), 'F##')
      assert.equal(pitch.str([0, 1, 3]), 'C#3')
    },
    'letter': function () {
      assert.equal(pitch.letter('fx'), 'F')
      assert.equal(pitch.letter('blah'), null)
    },
    'octave': function () {
      assert.equal(pitch.octave('C##4'), 4)
      assert.equal(pitch.octave([3, -2, 2]), 2)
      assert.equal(pitch.octave('C##'), null)
    },
    'pitch class': function () {
      assert.equal(pitch.pitchClass('e#2'), 'E#')
      assert.equal(pitch.pitchClass('dxx'), 'D####')
      assert.equal(pitch.pitchClass('db'), 'Db')
      assert.equal(pitch.pitchClass('blah'), null)
    },
    'accidentals': function () {
      assert.equal(pitch.accidentals('bbb3'), 'bb')
      assert.equal(pitch.accidentals('fx5'), '##')
      assert.equal(pitch.accidentals('c'), '')
      assert.equal(pitch.accidentals('blah'), null)
    }
  },
  'midi': {
    'fromMidi': function () {
      var notes = [60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73]
      .map(pitch.fromMidi).join(' ')
      assert.equal(notes, 'C4 Db4 D4 Eb4 E4 F4 F#4 G4 Ab4 A4 Bb4 B4 C5 Db5')
    },
    'toMidi': function () {
      assert.deepEqual('C4 D4 E4 F4 G4 A4 B4 C4'.split(' ').map(pitch.toMidi), [60, 62, 64, 65, 67, 69, 71, 60])
      assert.deepEqual('C#4 D#4 E#4 F#4 G#4 A#4 B#4 C#4'.split(' ').map(pitch.toMidi), [61, 63, 65, 66, 68, 70, 72, 61])
      assert.deepEqual('C##4 D##4 E##4 F##4 G##4 A##4 B##4 C##4'.split(' ').map(pitch.toMidi), [62, 64, 66, 67, 69, 71, 73, 62])
      assert.deepEqual('Cb4 Db4 Eb4 Fb4 Gb4 Ab4 Bb4 Cb4'.split(' ').map(pitch.toMidi), [59, 61, 63, 64, 66, 68, 70, 59])
      assert.deepEqual('Cbb3 Dbb3 Ebb3 Fbb3 Gbb3 Abb3 Bbb3 Cbb3'.split(' ').map(pitch.toMidi), [46, 48, 50, 51, 53, 55, 57, 46])
    },
    'toMidi: pitch class does not have midi': function () {
      assert.deepEqual('C D E F G A B C'.split(' ').map(pitch.toMidi), [ null, null, null, null, null, null, null, null ])
    },
    'toMidi: enharmonics': function () {
      assert.equal(pitch.toMidi('B#3'), pitch.toMidi('C4'))
      assert.equal(pitch.toMidi('B##3'), pitch.toMidi('Db4'))
    }
  },
  'frequencies': {
    'fromFreq': function () {
      assert.equal(pitch.fromFreq(440), 'A4')
      assert.equal(pitch.fromFreq(220), 'A3')
      assert.equal(pitch.fromFreq(329.6275569128699), 'E4')
      assert.equal(pitch.fromFreq(330), 'E4')
      assert.equal(pitch.fromFreq(335), 'E4')
      assert.equal(pitch.fromFreq(340), 'F4')
      assert.equal(pitch.fromFreq(349.2282314330039), 'F4')
    },
    'fromFreq with custom tuning': function () {
      assert.equal(pitch.fromFreq(220, 220), 'A4')
    },
    'toFreq': function () {
      assert.equal(pitch.toFreq('A4'), 440)
      assert.equal(pitch.toFreq('A3'), 220)
      assert.equal(pitch.toFreq('E4'), 329.6275569128699)
      assert.equal(pitch.toFreq('F4'), 349.2282314330039)
    },
    'toFreq custom tuning': function () {
      assert.equal(pitch.toFreq('A4', 444), 444)
      assert.equal(pitch.toFreq('A3', 444), 222)
    },
    'cents': function () {
      assert.equal(pitch.cents('A4', 'A#4'), 100)
      assert.equal(pitch.cents('A4', 444), 15.66)
    }
  }
}).export(module)


test/type-test.js
var vows = require('vows')
var assert = require('assert')
var type = require('../type')

var noop = function (x) { return x }
var up = function (x) { return x.toUpperCase() }
var rev = function (x) { return x.substring(1) }

vows.describe('type').addBatch({
  'simple': function () {
    var fn = type(up, rev)(noop)
    assert.equal(fn('ab'), 'B')
  }
}).export(module)


test/a-pitch-tests.js
var vows = require('vows')
var assert = require('assert')
var arr = require('../a-pitch')

var noop = function (x) { return x }
var up = function (x) { return x.toUpperCase() }
var rev = function (x) { return x.substring(1) }
var num = function (x) { return /^\d+$/.test(x) ? +x : null }

vows.describe('fn').addBatch({
  'simple': function () {
    var fn = arr.fn([up], rev, noop)
    assert.equal(fn('ab'), 'B')
  },
  'exclude nulls': function () {
    var add = function (n) { return n + 10 }
    var fn = arr.fn([num], null, add)
    assert.equal(fn(10), 20)
    assert.equal(fn('a'), null)
  }
}).export(module)


a-pitch.js

var identity = function (x) { return x }
var isArray = Array.isArray
var anything = { parse: identity, build: identity }

function type (parse, build) {
  if (arguments.length === 1) return type(parse.parse, parse.stringify)

  var cached = function (src) {
    if (typeof src === 'string') return parse(src)
    else if (isArray(src)) return build(src)
    else return null
  }
  cached.parse = function (value) {
    return isArray(value) ? value : cached(value)
  }
  cached.build = function (value) {
    return isArray(value) ? cached(value) : null
  }
  return cached
}

function fn (parsers, builder, fn) {
  var len = parsers.length
  builder = builder || identity

  return function () {
    var value
    var args = []
    for (var i = 0; i < len; i++) {
      value = parse(parsers[i], arguments[i])
      if (value === null) return
      args.push(value)
    }
    return builder(fn.apply(null, args))
  }
}

function parse (parser, value) {
  if (!(typeof value === 'string')) return value
  return (parser || identity)(value)
}

module.exports = {
  type: type,
  fn: fn,
  anything: anything
}


index.js
'use strict'

var asPitch = require('pitch-parser')

// return the pitch as array
function asArray (pitch) {
  return Array.isArray(pitch) ? pitch : asPitch.parse(pitch)
}

// decorate a function to force the param to be an array-pitch
function parseDecorator (fn) {
  return function (pitch) {
    var p = asArray(pitch)
    return p ? fn(p) : null
  }
}

// Semitones from C to C D E F G A B
var SEMITONES = [ 0, 2, 4, 5, 7, 9, 11 ]
// Chromatic melodic scale
var CHROMATIC = [ 'C', 'Db', 'D', 'Eb', 'E', 'F', 'F#', 'G', 'Ab', 'A', 'Bb', 'B' ]

var lib = asArray
module.exports = lib

/**
 * Get pitch name in scientific notation
 *
 * @param {String|Array} pitch - the pitch string or array
 * @return {String} the name of the pitch
 *
 * @example
 * pitch.str('cb') // => 'Cb'
 * pitch.str('fx2') // => 'F##2'
 */
function str (pitch) {
  var p = Array.isArray(pitch) ? pitch : asPitch.parse(pitch)
  return p ? asPitch.stringify(p) : null
}
lib.str = str

/**
 * Get letter of the pitch (in uppercase)
 *
 * @param {String|Array} pitch - the pitch as string or array
 * @return {String} the letter of the pitch in uppercase
 *
 * @example
 * pitch.letter('fx') // => 'F'
 */
function letter (pitch) {
  var n = str(pitch)
  return n ? n.slice(0, 1) : null
}
lib.letter = letter

/**
 * Get the octave of a pitch
 *
 * @param {String|Array} pitch - the pitch
 * @return {Integer} the octave number
 *
 * @example
 * pitch.octave('F#3') // => 3
 */
function octave (pitch) {
  return pitch[2]
}
lib.octave = parseDecorator(octave)

/**
 * Get the pitch class (pitch name without octaves) from a pitch
 *
 * @param {String|Array} pitch - the pitch to get the pitchClass number from
 * @return {String} the pitch class
 *
 * @example
 * pitch.pitchClass('a4') // => 'A'
 * pitch.pitchClass('ab') // => 'Ab'
 * pitch.pitchClass('cx2') // => 'C##'
 */
function pitchClass (p) {
  if (!p) return null
  else if (Array.isArray(p)) return [p[0], p[1], null]
  p = asPitch.parse(p)
  return p ? asPitch.stringify(pitchClass(p)) : null
}
lib.pitchClass = pitchClass

/**
 * Get the accidentals from a pitch
 *
 * @param {String|Array} pitch - the pitch
 * @return {String} the pitch accidentals
 *
 * @example
 * pitch.accidentals('C##3') // => '##'
 * pitch.accidentals('Bb4') // => 'b'
 * pitch.accidentals('E') // => ''
 */
function accidentals (p) {
  var pc = pitchClass(p)
  return pc ? pc.substring(1) : null
}
lib.accidentals = accidentals

/**
 * Get the pitch of the given midi number
 *
 * This method doesn't take into account diatonic spelling. Always the same
 * pitch class is given to the same midi number.
 *
 * @param {Integer} midi - the midi number
 * @return {String} the pitch
 *
 * @example
 * pitch.fromMidi(69) // => 'A4'
 */
function fromMidi (midi) {
  var name = CHROMATIC[midi % 12]
  var oct = Math.floor(midi / 12) - 1
  return name + oct
}
lib.fromMidi = fromMidi

/**
 * Get the midi number of a pitch
 *
 * @param {String|Array} pitch - the pitch string (or pitch array)
 * @return {Integer} the midi number
 *
 * @example
 * pitch.toMidi('A4') // => 69
 * pitch.toMidi('A3') // => 57
 */
function toMidi (p) {
  if (!p[2] && p[2] !== 0) return null
  return SEMITONES[p[0]] + p[1] + 12 * (p[2] + 1)
}
lib.toMidi = parseDecorator(toMidi)

/**
 * Get the pitch of a given frequency.
 *
 * It will round the frequency to the nearest pitch frequency. Use `cents` function
 * if you need to know the difference between the the frequency and the pitch.
 *
 * @param {Float} freq - the frequency
 * @return {String} the pitch
 *
 * @see cents
 *
 * @example
 * pitch.fromFreq(440) // => 'A4'
 * pitch.fromFreq(443) // => 'A4'
 * pitch.cents(443, 'A4') // => ... to get the difference
 */
function fromFreq (freq, tuning) {
  tuning = tuning || 440
  var lineal = 12 * ((Math.log(freq) - Math.log(tuning)) / Math.log(2))
  var midi = Math.round(69 + lineal)
  return fromMidi(midi)
}
lib.fromFreq = fromFreq

// decimal number
var NUM = /^\d+(?:\.\d+)?$/
/**
 * Get the pitch frequency in hertzs
 *
 * @param {String} pitch - the pitch
 * @param {Integer} tuning - optional tuning, 440 by default
 * @return {Float} - the pitch frequency
 *
 * @example
 * pitch.toFreq('A4') // => 440
 * pitch.toFreq('A3', 444) // => 222
 */
function toFreq (p, tuning) {
  if (NUM.test(p)) return +p
  var midi = toMidi(asPitch.parse(p))
  if (!midi) return null
  tuning = tuning || 440
  return Math.pow(2, (midi - 69) / 12) * tuning
}
lib.toFreq = toFreq

/**
 * Get the distance in cents between pitches or frequencies
 *
 * @param {String|Integer} from - first pitch or frequency
 * @param {String|Integer} to - other pitch or frequency
 * @param {Integer} decimals - the decimal precision (2 by default)
 * @return {Integer} the distance in cents
 *
 * @example
 * cents(440, 444) // => 15.66
 * cents('A4', 444) // => 15.66
 * cents('A4', 'A#4') // => 100
 */
function cents (from, to, decimals) {
  var dec = decimals ? Math.pow(10, decimals) : 100
  var fromFq = toFreq(from)
  var toFq = toFreq(to)
  return Math.floor(1200 * (Math.log(toFq / fromFq) * dec / Math.log(2))) / dec
}
lib.cents = cents
